### PR TITLE
Add support for special hours

### PR DESCRIPTION
**Pull request by meganle**
My Python is pretty ugly so plz help

- Scrapes special hours from this page https://recreation.athletics.cornell.edu/hours-facilities/cornell-fitness-center-special-hours
- Checks if special hours apply in the next 7 days and updates hours accordingly
- Added `specialHours` field for debugging purposes
- Screenshot below is from querying the hours for Appel. It includes the special hours from 3/1 (today) until 3/3 (when special hours end) and also includes the regular hours for the remaining days
- If a gym is closed for a certain day, the "times" array won't have an entry for that day (Not sure how iOS would handle gyms that are closed for the entire day)

screen shot 2019-03-01 at 4 05 41 pm

### CODE REVIEW

File: src/data.py [hunk 1]
```diff
@@ -7,10 +8,58 @@
...
     class_details, classes = scraper.scrape_classes(constants.PAGE_LIMIT)
     Data.update_data(gyms=gyms, classes=classes, class_details=class_details, limit=constants.CLASS_HISTORY_LIMIT)
   finally:
     threading.Timer(constants.UPDATE_DELAY, start_update).start()
 
+'''
+Check for special hours and update hours if applicable
+Returns:
+  list of GymType objects with special hours if applicable else uses regular hours
+'''
```
Comment (young-k) [inline src/data.py]: docstrings for python go inside of the function, and are denoted by `"""` instead of `'''`

File: src/data.py [hunk 2]
```diff
@@ -7,10 +8,58 @@
...
+  if any(special_hours):
+    for gym in gyms:
+      if 'Teagle' in gym.name:
+        # Since we have both Teagle Up and Teagle Down
+        times = special_hours['Teagle Hall']
+      else:
+        times = special_hours[gym.name]
+      
+      start_index = None
+      for i in range(0, len(times)):
```
Comment (young-k) [inline src/data.py]: `for i in range(len(times))`

File: src/scraper.py [hunk 3]
```diff
@@ -129,3 +140,81 @@ def get_image_url(name):
...
   return constants.ASSET_BASE_URL + 'classes/' + image_keyword + '.jpg'
+
+'''
+Scrape special gym hours
+Returns:
+  dict mapping gym names to list of dicts each with a date and a DayTimeRangeType object
+
+Example:
+  { 'Helen Newman': [ {'date': '2/18', 'day': 1, 'hours': <DayTimeRangeType object>}, ...] }
+'''
```
Comment (young-k) [inline src/scraper.py]: same comment here

File: src/scraper.py [hunk 4]
```diff
@@ -129,3 +140,81 @@ def get_image_url(name):
...
+          mid_index = time_text.index('-')
+
+          # The start and end times use 'a' or 'p' to indicate 'am' vs 'pm'
+          start_time_string = time_text[0:mid_index] + 'm'
+          end_time_string = time_text[mid_index+1:] + 'm'
+
+          start_time = datetime.strptime(start_time_string, '%I:%M%p').time()
+          end_time = datetime.strptime(end_time_string, '%I:%M%p').time()
+
+          for i in range(0, days):
```
Comment (young-k) [inline src/scraper.py]: `range(days)`

File: src/scraper.py [hunk 5]
```diff
@@ -129,3 +140,81 @@ def get_image_url(name):
...
+      current_date_index = 0
+
+      for time in times:
+        time_text = time.text.strip().replace('Noon', '12:00p')
+
+        # Number of consecutive days these hours are effective
+        days = int(time.get('colspan', 1))
+
+        if time_text == 'Closed':
+          for i in range(0, days):
```
Comment (young-k) [inline src/scraper.py]: `range(days)`

File: src/schema.py [hunk 6]
```diff
@@ -29,6 +29,7 @@ class DayTimeRangeType(ObjectType):
   day = Int(required=True)
   start_time = Time(required=True)
   end_time = Time(required=True)
+  special_hours = Boolean(default_value=False, required=True)
```
Comment (young-k) [inline src/schema.py]: these should probably be alphabetized

File: src/data.py [hunk 7]
```diff
@@ -7,10 +8,58 @@
...
+  except:
+    special_hours = None
+
+  # Deep copy so the regular gym times aren't modified
+  gyms = deepcopy(constants.GYMS)
+
+  if any(special_hours):
+    for gym in gyms:
+      if 'Teagle' in gym.name:
+        # Since we have both Teagle Up and Teagle Down
```
Comment (young-k) [inline src/data.py]: maybe move this comment to the previous line, up to you though

File: src/data.py (hunk 2 above)
Comment (williambaibai) [reply src/data.py]: +1

File: src/data.py [hunk 8]
```diff
@@ -7,10 +8,58 @@
...
+        if times[i]['date'] == current_date:
+          start_index = i
+          break
+
+      if not start_index:
+        # No special hours for gym
+        continue
+
+      end_index = min(start_index + 7, len(times))
+      hours = [e['hours'] for e in times[start_index:end_index] if e['hours']]
```
Comment (williambaibai) [inline src/data.py]: maybe a descriptive variable than `e`? Such as `elt` or `gym_time`?

File: src/data.py [hunk 9]
```diff
@@ -7,10 +8,58 @@
...
+      # Indices of days given by special hours
+      days_covered = [e['day'] for e in times[start_index:end_index]]
```
Comment (williambaibai) [inline src/data.py]: Same here ^^

File: src/schema.py (hunk 6 above)
Comment (williambaibai) [reply src/schema.py]: +1

File: src/scraper.py [hunk 10]
```diff
@@ -129,3 +140,81 @@ def get_image_url(name):
...
+'''
+def scrape_special_hours():
+  page = requests.get(BASE_URL + SPECIAL_HOURS_PATH).text
+  soup = BeautifulSoup(page, 'lxml')
+  schedules = soup.find_all('table')
+  gym_hours = {}
+
+  for table in schedules:
+    rows = table.find_all('tr')
+
```
Comment (williambaibai) [inline src/scraper.py]: Is the new line here needed?

File: src/data.py [hunk 11]
```diff
@@ -7,10 +8,58 @@
...
   finally:
     threading.Timer(constants.UPDATE_DELAY, start_update).start()
 
+'''
+Check for special hours and update hours if applicable
+Returns:
+  list of GymType objects with special hours if applicable else uses regular hours
+'''
+def check_for_special_hours():
+  current_date = '{0}/{1}'.format(datetime.now().month, datetime.now().day)
```
Comment (kevinchan159) [inline src/data.py]: maybe do 
`now = datetime.now()`
`current_date = '{0}/{1}'.format(now.month, now.day)`
not sure if its better

**Pull request merged**